timidity.1
.TH TiMidity 1 "8 Sep 1995" \" -*-nroff-*-
.SH NAME
TiMidity \- MIDI to WAVE converter and player
.P
.SH SYNOPSIS
.B timidity
[\-options] filename [filenames ...]
.P
.SH DESCRIPTION
TiMidity is a MIDI to WAVE converter using Gravis
Ultrasound\-compatible patch files to generate digital audio data from
General MIDI files.  The data can be stored in a file for processing,
or played in real time through an audio device.
.P
.SH OPTIONS
The following command line options are accepted by version 0.2i of
TiMidity:
.TP
.BI \-o " filename"
Place output on \fIfilename\fP, which may be a file, device, or HP-UX
audio server, depending on the output mode selected with the \fB\-O\fP
option. The special filename ``\-'' causes output to be placed on
stdout.
.TP
.BI \-O " mode"
Selects the output mode from the compiled-in alternatives.  \fImode\fP
must begin with one of the supported output mode identifiers.  Run
TiMidity with the \fB\-h\fP option to see a list.  The following
identifiers should be available in all versions:
.RS
.TP
.B \-Or
Generate raw waveform data.  All format options are supported.  Common
formats include:
.RS
.TP
.B \-OrU
uLaw
.TP
.B \-Or1sl
16-bit signed linear PCM
.TP
.B \-Or8ul
8-bit unsigned linear PCM
.RE
.TP
.B \-Ow
Generate RIFF WAVE format output.  If output is directed to a
non-seekable file, or if TiMidity is interrupted before closing the
file, the file header will contain 0xFFFFFFFF in the RIFF and data
block length fields.  The popular sound conversion utility \fIsox\fP
is able to read such malformed files, so you can pipe data directly to
sox for on\-the\-fly conversion to other formats.
.TP
.B "Format options"
Option characters may be added immediately after the mode identifier
to change the output format.  The following options are recognized:
.RS
.LP
.TP
.B 8
8-bit sample width
.TP
.B 1
16-bit sample width
.TP
.B l
Linear encoding
.TP
.B U
uLaw (8-bit) encoding
.TP
.B M
Monophonic
.TP
.B S
Stereo
.TP
.B s
Signed output
.TP
.B u
Unsigned output
.TP
.B x
Byte-swapped output
.RE
.LP
Note that some options have no effect on some modes.  For example, you
cannot generate a byte-swapped RIFF WAVE file, or force uLaw output on
a Linux PCM device.
.RE
.LP
.TP
.BI \-s " frequency"
Sets the resampling frequency.  Not all sound devices are capable of
all frequencies -- an approximate frequency may be selected, depending
on the implementation.
.TP
.B \-a
Turns on antialiasing. Samples are run through a lowpass filter before
playing, which reduces aliasing noise at low resampling frequencies.
.TP
.B \-f
Toggles fast envelopes. Some MIDI files sound better when notes decay
slower -- it gives the impression of reverb, which TiMidity doesn't
currently support.
.TP
.B \-e
Make TiMidity evil.  On Win32 version, this increases the task
priority by one. It can give better playback when you switch tasks at
the expense of slowing all other tasks down.
.TP
.BI \-p " voices"
Sets polyphony (maximum number of simultaneous voices) to
\fIvoices\fP.
.TP
.BI \-A " amplification"
Multiplies the master volume by \fIamplification\fP%.
.TP
.BI \-C " ratio"
Sets the ratio of sampling and control frequencies.  This determines how
often envelopes are recalculated -- small ratios yield better quality
but use more CPU time.
.TP
.BI \-L " directory"
Adds \fIdirectory\fP to the library path.  Patch, configuration, and
MIDI files are searched along this path.  Directories added last will
be searched first.  Note that the current directory is always searched
first before the library path.
.TP
.BI \-c " file"
Reads an extra configuration file.
.TP
.BI \-I " number"
Uses the program \fInumber\fP as the default instrument.  Any Program
Change events in MIDI files will override this option.
.TP
.BI \-P " file"
Uses the patch \fIfile\fP for every program except drums.  Program
Change events will be ignored.  This is useful for testing new
patches.
.TP
.BI \-D " channel"
Marks \fIchannel\fP as a drum channel.  
If \fIchannel\fP is negative, channel \-\fIchannel\fP is marked as an
instrumental channel.
If \fIchannel\fP is 0, all channels are marked as instrumental.
.TP 
.BI \-Q " channel"
Causes \fIchannel\fP to be quiet. 
If \fIchannel\fP is negative, channel \-\fIchannel\fP is turned back on. 
If \fIchannel\fP is 0, all channels are turned on.
.TP 
.B \-F
Turns on fast panning to accommodate MIDI pieces that expect panning
adjustments to affect notes that are already playing.  Some files that
don't expect this have a habit of flipping balance rapidly between
left and right, which can cause severe popping when the -F flag is
used.
.TP
.BI \-U
Instructs TiMidity to unload all instruments from memory between
MIDI files.  This can reduce memory requirements when playing many
files in succession.
.TP
.BI \-i " interface"
Selects the user interface from the compiled-in alternatives.
\fIinterface\fP must begin with one of the supported interface
identifiers.  Run TiMidity with the \fB\-h\fP option to see a list.
The following identifiers may be available:
.RS
.TP
.B \-id
The dumb interface -- plays files in sequence, prints messages
according to verbosity level.  The trace mode shows the current and
total playing time.
.TP
.B \-im
The motif interface -- X Window interactive interface.
.TP
.B \-ik
The Tcl/Tk interface -- X Window interactive interface.
.TP
.B \-in
The ncurses full\-screen interface with interactive controls.
.TP
.B \-is
The S-Lang full\-screen interface with interactive controls.
.TP
.B "Interface options"
Option characters may be added immediately after the interface
identifier.  The following options are recognized:
.RS
.TP
.B v
Increases verbosity.  This option is cumulative.
.TP
.B q
Decreases verbosity.  This option is cumulative.
.TP
.B t
Toggles trace mode.  In trace mode, TiMidity attempts to display its
current state in real time.  For the Linux sound driver, this is
accomplished through the use of short DMA buffer fragments, which can
be tuned via the \fB\-B\fP option.
.RE
.RE
.LP
.TP
.TP
.BI \-B " fragments"
For the Linux sound driver, selects the number of buffer fragments in
interactive mode.  Increasing the number of fragments may reduce
choppiness when many processes are running.  It will make TiMidity seem
to respond sluggishly to fast forward, rewind, and volume controls,
and it will throw the status display off sync.  Specify a
\fIfragments\fP of 0 to use the maximum number of fragments available.
.P
.SH FILES
TiMidity looks for the configuration file \fItimidity.cfg\fP at
startup, before processing any options.  If it can't be accessed, and
the library path is changed with a \fB\-L\fP option on the command
line, then the default file will be sought again along the new library
path after processing all options, unless another configuration file was
specified with the \fB\-c\fP option. 
.P
Configuration files define the mapping of MIDI programs to instrument
files.  Multiple files may be specified, and statements in later ones
will override earlier ones.  The following statements can be used in a
configuration file:
.TP
.BI dir " directory"
Adds \fIdirectory\fP to the search path in the same manner as the
\fB\-L\fP command line option.
.TP
.BI source " file"
Reads another configuration file, then continues processing the
current one.
.TP
.BI bank " number"
Selects the tone bank to modify.  Patch mappings that follow will
affect this tone bank.
.TP
.BI drumset " number"
Selects the drum set to modify.  Patch mappings that follow will affect
this drum set.
.TP
.I "number file [options]"
Specifies that the the MIDI program \fInumber\fP in the current tone
bank or drum set should be played using the patch \fIfile\fP.
\fIoptions\fP may be any of the following:
.RS
.TP
\fBamp=\fP\fIamplification\fP
Amplifies the instrument's volume by \fIamplification\fP percent.
If no value is specified, one will be automatically determined whenever
the instrument is loaded.
.TP
\fBnote=\fP\fInote\fP
Specifies a fixed MIDI note to use when playing the instrument.
If \fInote\fP is 0, the instrument will be played at whatever note
the Note On event triggering it has. For percussion instruments, if no
value is specified in the configuration file, the default in the patch
file will be used.
.TP
\fBpan=\fP\fIpanning\fP
Sets the instrument's default panning. 
\fIpanning\fP may be \fBleft\fP, \fBright\fP, \fBcenter\fP, or an
integer between -100 and 100, designating full left and full right
respectively. 
If no value is specified, the default in the patch file will be used. 
Note that panning controls in MIDI files will override this value.
.TP
\fBkeep=\fP{\fBloop\fP|\fBenv\fP}
By default, percussion instruments have their loop and envelope
information stripped. Strangely shaped envelopes are removed
automatically from melodic instruments as well. \fBkeep\fP can be used
to prevent stripping envelope or loop data. For example, the Short and
Long Whistle percussion instruments (General Midi numbers 71 and 72)

NEWS
11/24, 1999
    * Version 2.8.0 released.
    * In GTK, added support for printing the lyrics from .kar files
      and also make the text area clear between files.
    * New command line option: -M <name>
	-M auto: Use *.mid.wav or *.mid.aiff instread of MIDI to WAVE
		 convertion.  The PCM file format must be 16bit sample and
		 the sampling rate equals as timidity's rate.
        -M none: Disable this feature (default)
        -M <filename>: Use the specified file.
    * In GTK interface, added support for printing the lyrics from .kar files
      and also make the text area clear between files.
    * New configure option to choose wave interpolation in resampling.
	--with-spline=linear	Linear interpolation (default)
	--with-spline=cubic	Cubic interpolation
	--with-spline=lanrange	Langrange interpolation
	--with-spline=no	No interplation
    * New command line option `-m N' to specify sustain time (msec).
	-m 1	Effectively behaves as if all sustains are ignored,
		volume ramping is the same as normal stage 3.
	-m 0	Disable sustain ramping, causes constant volume sustains.
		(default)
	-m 3000	A note at full volume will decay for 3 seconds once it begins
		to be sustained (assuming the regular stage 3 rate would not
		cause it to decay even longer). Softer notes will of course
		die sooner.

10/12, 1999
    * Version 2.7.0 released.
    * adjust_panning_immediately (-F) is true by default.
    * GS Sysex for bank+program change.
    * Bug fix.

9/27, 1999
    * Version 2.6.1 released.
    * Support for XG Multi Part Data parameter change SysEx events.
    * Added new Channel structure member `int pan_random' for random pan flag.
      Note that all of dynamic interfaces must be re-compiled,
      because the size of `Channel' structure is changed.
    * Bug fix.

9/21, 1999
    * Version 2.6.0 released.
    * Some configurations and installations are changed.
    * -EFchorus=2 for the surround sound instead of detuned chorus.
    * Bug fix.

8/27, 1999
    * Version 2.5.1 released.
    * Bug fix.

8/26, 1999
    * Version 2.5.0 released.
    * Upgrade Windows GUI.
    * MIMPI WRD for TrueColor X Window works more stable.
    * In ncurses interface, support to display GSLCD.
    * In ncurses interface, Enable to display channels more than 16ch.
      The terminal must be able to display more than 24 rows to display
      more than 16ch.
    * On Windows, recognize both `\' and `/' as path separator if the
      binary is compiled with Cygwin.
    * Improvement for CSPLINE_INTERPOLATION and LAGRANGE_INTERPOLATION.
    * Less CPU% in trace playing.
    * Bug fix.

8/11, 1999
    * Version 2.4.1 released.
    * Bug fix.

8/9, 1999
    * Version 2.4.0 released.
    * Make archive library (libarc) code more simple.
    * On linux, fixed problem of audio queue analysing.

8/1, 1999
    * Version 2.3.0 released.
    * Make Windows GUI code elegant.
    * Support MIDI to WAVE converter in Windows GUI.

7/26, 1999
    * New feature: Auto polyphony reduction.

7/19, 1999
    * Version 2.2.0 released.
    * Enabled to specify the software buffer size with `-q' option
    * Enabled build UMP on FreeBSD-ELF environment.

7/9, 1999
    * Version 2.1.1 released.

6/10, 1999
    * Version 2.1.0 released.
    * Enable a multi-point interpolation in resampling.
      Cubic spline curve and lagrange interpolation is supported.
    * Support linux axp
    * Load frequency table with -Z <file> option.

4/28, 1999
    * Version 2.0.2 released.

4/25, 1999
    * Version 2.0.1 released.

4/17, 1999
Major changes in release 2.0.0:

* Auto voice reduction
  If there is no enough machine power to play the MIDI in realtime,
  TiMidity automatically reduces the voice.
* Audio buffering
  TiMidity has internal audio buffer.  TiMidity doesn't start to play until
  this buffer is filled.
* Support Windows GUI.
* Support Sherry WRD.
* Support NAS (Network Audio System) output.
* Support EsounD output.
* Enable to play the stream MIDI in realtime.  The stream format is same
  as OSS /dev/sequencer stream.  See the protocol in interface/server_c.c
* Enable to assign patch from part of soundfont.
* Lower CPU load average in trace playing.
  v1 is implemented with busy-wait-method to synchronize audio and
  interface, but v2 is not.
* Reverb and chorus effect is enabled by default.  You can disable their
  controls with command line option.

timidity.c
/*

    TiMidity -- Experimental MIDI to WAVE converter
    Copyright (C) 1995 Tuukka Toivonen <toivonen@clinet.fi>

	 This program is free software; you can redistribute it and/or modify
	 it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
	 (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
	 GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

*/
#include <stdio.h>
#include <stdlib.h>

#if defined(SOLARIS) || defined(__WIN32__)
#include <string.h>
#else
#include <strings.h>
#endif

#ifdef __WIN32__
#include <windows.h>
#define rindex strrchr
extern int optind;
extern char *optarg;
int getopt(int, char **, char *);
#else
#include <unistd.h>
#endif

#include "config.h"
#include "common.h"
#include "instrum.h"
#include "playmidi.h"
#include "readmidi.h"
#include "output.h"
#include "controls.h"
#include "tables.h"

int free_instruments_afterwards=0;
static char def_instr_name[256]="";

#ifdef __WIN32__
int intr;
CRITICAL_SECTION critSect;

#pragma argsused
static BOOL WINAPI handler (DWORD dw)
	{
	printf ("***BREAK\n");
	intr = TRUE;
	play_mode->purge_output ();
	return TRUE;
	}
#endif

static void help(void)
{
  PlayMode **pmp=play_mode_list;
  ControlMode **cmp=ctl_list;
  printf(" TiMidity version " TIMID_VERSION " (C) 1995 Tuukka Toivonen "
	 "<toivonen@clinet.fi>\n"
	 " TiMidity is free software and comes with ABSOLUTELY NO WARRANTY.\n"
	 "\n"
#ifdef __WIN32__
	 " Win32 version by Davide Moretti <dmoretti@iper.net>\n"
	 "\n"
#endif
	 "Usage:\n"
	 "  %s [options] filename [...]\n"
	 "\n"
#ifndef __WIN32__		/*does not work in Win32 */
	 "  Use \"-\" as filename to read a MIDI file from stdin\n"
	 "\n"
#endif
	 "Options:\n"
#if defined(AU_HPUX)
	 "  -o file Output to another file (or audio server) (Use \"-\" for stdout)\n"
#elif defined (AU_LINUX)
	 "  -o file Output to another file (or device) (Use \"-\" for stdout)\n"
#else
	 "  -o file Output to another file (Use \"-\" for stdout)\n"
#endif
	 "  -O mode Select output mode and format (see below for list)\n"
	 "  -s f    Set sampling frequency to f (Hz or kHz)\n"
	 "  -a      Enable the antialiasing filter\n"
	 "  -f      "
#ifdef FAST_DECAY
					"Disable"
#else
					"Enable"
#endif
							  " fast decay mode\n"
	 "  -p n    Allow n-voice polyphony\n"
	 "  -A n    Amplify volume by n percent (may cause clipping)\n"
	 "  -C n    Set ratio of sampling and control frequencies\n"
	 "\n"
	 "  -L dir  Append dir to search path\n"
	 "  -c file Read extra configuration file\n"
	 "  -I n    Use program n as the default\n"
	 "  -P file Use patch file for all programs\n"
	 "  -D n    Play drums on channel n\n"
	 "  -Q n    Ignore channel n\n"
	 "  -F      Enable fast panning\n"
	 "  -U      Unload instruments from memory between MIDI files\n"
	 "\n"
	 "  -i mode Select user interface (see below for list)\n"
#if defined(AU_LINUX) || defined(AU_WIN32)
	 "  -B n    Set number of buffer fragments\n"
#endif
#ifdef __WIN32__
	 "  -e      Increase thread priority (evil) - be careful!\n"
#endif
	 "  -h      Display this help message\n"
	 "\n"
	 , program_name);

  printf("Available output modes (-O option):\n\n");
  while (*pmp)
	 {
		printf("  -O%c     %s\n", (*pmp)->id_character, (*pmp)->id_name);
		pmp++;
	 }
  printf("\nOutput format options (append to -O? option):\n\n"
	 "   `8'    8-bit sample width\n"
	 "   `1'    16-bit sample width\n"
	 "   `U'    uLaw encoding\n"
	 "   `l'    linear encoding\n"
	 "   `M'    monophonic\n"
	 "   `S'    stereo\n"
	 "   `s'    signed output\n"
	 "   `u'    unsigned output\n"
	 "   `x'    byte-swapped output\n");

  printf("\nAvailable interfaces (-i option):\n\n");
  while (*cmp)
	 {
      printf("  -i%c     %s\n", (*cmp)->id_character, (*cmp)->id_name);
      cmp++;
	 }
  printf("\nInterface options (append to -i? option):\n\n"
	 "   `v'    more verbose (cumulative)\n"
	 "   `q'    quieter (cumulative)\n"
	 "   `t'    trace playing\n\n");
}

static void interesting_message(void)
{
  printf(
"\n"
" TiMidity version " TIMID_VERSION " -- Experimental MIDI to WAVE converter\n"
" Copyright (C) 1995 Tuukka Toivonen <toivonen@clinet.fi>\n"
" \n"
#ifdef __WIN32__
" Win32 version by Davide Moretti <dmoretti@iper.net>\n"
"\n"
#endif
" This program is free software; you can redistribute it and/or modify\n"
" it under the terms of the GNU General Public License as published by\n"
" the Free Software Foundation; either version 2 of the License, or\n"
" (at your option) any later version.\n"
" \n"
" This program is distributed in the hope that it will be useful,\n"
" but WITHOUT ANY WARRANTY; without even the implied warranty of\n"
" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the\n"
" GNU General Public License for more details.\n"
" \n"
" You should have received a copy of the GNU General Public License\n"
" along with this program; if not, write to the Free Software\n"
" Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.\n"
"\n"
);

}

static int set_channel_flag(int32 *flags, int32 i, char *name)
{
  if (i==0) *flags=0;
  else if ((i<1 || i>16) && (i<-16 || i>-1))
    {
		fprintf(stderr,
	      "%s must be between 1 and 16, or between -1 and -16, or 0\n", 
	      name);
      return -1;
	 }
  else
    {
      if (i>0) *flags |= (1<<(i-1));
		else *flags &= ~ ((1<<(-1-i)));
    }
  return 0;
}

static int set_value(int32 *param, int32 i, int32 low, int32 high, char *name)
{
  if (i<low || i > high)
    {
      fprintf(stderr, "%s must be between %ld and %ld\n", name, low, high);
      return -1;
    }
  else *param=i;
  return 0;
}

static int set_play_mode(char *cp)
{
  PlayMode *pmp, **pmpp=play_mode_list;

  while ((pmp=*pmpp++))
    {
      if (pmp->id_character == *cp)
	{
	  play_mode=pmp;
	  while (*(++cp))
		 switch(*cp)
	      {
	      case 'U': pmp->encoding |= PE_ULAW; break; /* uLaw */
			case 'l': pmp->encoding &= ~PE_ULAW; break; /* linear */

			case '1': pmp->encoding |= PE_16BIT; break; /* 1 for 16-bit */
	      case '8': pmp->encoding &= ~PE_16BIT; break;

	      case 'M': pmp->encoding |= PE_MONO; break;
	      case 'S': pmp->encoding &= ~PE_MONO; break; /* stereo */

	      case 's': pmp->encoding |= PE_SIGNED; break;
	      case 'u': pmp->encoding &= ~PE_SIGNED; break;

	      case 'x': pmp->encoding ^= PE_BYTESWAP; break; /* toggle */

			default:
		fprintf(stderr, "Unknown format modifier `%c'\n", *cp);
		return 1;
	      }
	  return 0;
	}
	 }
  
  fprintf(stderr, "Playmode `%c' is not compiled in.\n", *cp);
  return 1;
}

static int set_ctl(char *cp)
{
  ControlMode *cmp, **cmpp=ctl_list;

  while ((cmp=*cmpp++))
    {
      if (cmp->id_character == *cp)
	{
	  ctl=cmp;
	  while (*(++cp))
	    switch(*cp)
			{
			case 'v': cmp->verbosity++; break;
	      case 'q': cmp->verbosity--; break;
			case 't': /* toggle */
		cmp->trace_playing= (cmp->trace_playing) ? 0 : 1; 
		break;

	      default:
		fprintf(stderr, "Unknown interface option `%c'\n", *cp);
		return 1;
	      }
	  return 0;
	}
	 }
  
  fprintf(stderr, "Interface `%c' is not compiled in.\n", *cp);
  return 1;
}

#define MAXWORDS 10

static int read_config_file(char *name)
{
  FILE *fp;
  char tmp[1024], *w[MAXWORDS], *cp;
  ToneBank *bank=0;
  int i, j, k, line=0, words;
  static int rcf_count=0;

  if (rcf_count>50)
	 {
		fprintf(stderr, "Probable source loop in configuration files");
		return (-1);
	 }

  if (!(fp=open_file(name, 1, OF_VERBOSE)))
	 return -1;

  while (fgets(tmp, sizeof(tmp), fp))
	 {
      line++;
		w[words=0]=strtok(tmp, " \t\r\n\240");
		if (!w[0] || (*w[0]=='#')) continue;
      while (w[words] && (words < MAXWORDS))
	w[++words]=strtok(0," \t\r\n\240");
      if (!strcmp(w[0], "dir"))
	{
	  if (words < 2)
		 {
	      fprintf(stderr, "%s: line %d: No directory given\n", name, line);
	      return -2;
		 }
	  for (i=1; i<words; i++)
	    add_to_pathlist(w[i]);
	}
		else if (!strcmp(w[0], "source"))
	{
	  if (words < 2)
	    {
	      fprintf(stderr, "%s: line %d: No file name given\n", name, line);
	      return -2;
		 }
	  for (i=1; i<words; i++)
	    {
	      int status;
	      rcf_count++;
	      status = read_config_file(w[i]);
	      rcf_count--;
	      if(status != 0)
		return status;
	    }
	}
      else if (!strcmp(w[0], "default"))
	{
	  if (words != 2)
	    {
	      fprintf(stderr, 
				"%s: line %d: Must specify exactly one patch name\n",
		      name, line);
	      return -2;
	    }
	  strncpy(def_instr_name, w[1], 255);
	  def_instr_name[255]='\0';
	}
		else if (!strcmp(w[0], "drumset"))
	{
	  if (words < 2)
	    {
	      fprintf(stderr, "%s: line %d: No drum set number given\n", 
		      name, line);
			return -2;
	    }
	  i=atoi(w[1]);
	  if (i<0 || i>127)
		 {
	      fprintf(stderr, 
		      "%s: line %d: Drum set must be between 0 and 127\n",
				name, line);
	      return -2;
		 }
	  if (!drumset[i])
	    {
	      drumset[i]=safe_malloc(sizeof(ToneBank));
			memset(drumset[i], 0, sizeof(ToneBank));
		 }
	  bank=drumset[i];
	}
		else if (!strcmp(w[0], "bank"))
	{
	  if (words < 2)
		 {
	      fprintf(stderr, "%s: line %d: No bank number given\n", 
				name, line);
	      return -2;
		 }
	  i=atoi(w[1]);
	  if (i<0 || i>127)
	    {
	      fprintf(stderr, 
		      "%s: line %d: Tone bank must be between 0 and 127\n",
				name, line);
	      return -2;
	    }
	  if (!tonebank[i])
		 {
			tonebank[i]=safe_malloc(sizeof(ToneBank));
	      memset(tonebank[i], 0, sizeof(ToneBank));
	    }
	  bank=tonebank[i];
	}
      else {
	if ((words < 2) || (*w[0] < '0' || *w[0] > '9'))
	  {
		 fprintf(stderr, "%s: line %d: syntax error\n", name, line);
		 return -2;
	  }
	i=atoi(w[0]);
	if (i<0 || i>127)
	  {
	    fprintf(stderr, "%s: line %d: Program must be between 0 and 127\n",
		    name, line);
	    return -2;
	  }
	if (!bank)
	  {
	    fprintf(stderr, 
			 "%s: line %d: Must specify tone bank or drum set "
		    "before assignment\n",
		    name, line);
		 return -2;
	  }
	if (bank->tone[i].name)
	  free(bank->tone[i].name);
	strcpy((bank->tone[i].name=safe_malloc(strlen(w[1])+1)),w[1]);
	bank->tone[i].note=bank->tone[i].amp=bank->tone[i].pan=
	  bank->tone[i].strip_loop=bank->tone[i].strip_envelope=
	    bank->tone[i].strip_tail=-1;

	for (j=2; j<words; j++)
	  {
	    if (!(cp=strchr(w[j], '=')))
	      {
		fprintf(stderr, "%s: line %d: bad patch option %s\n",
			name, line, w[j]);
		return -2;
	      }
	    *cp++=0;
	    if (!strcmp(w[j], "amp"))
			{
		k=atoi(cp);
		if ((k<0 || k>MAX_AMPLIFICATION) || (*cp < '0' || *cp > '9'))
		  {
			 fprintf(stderr,
			    "%s: line %d: amplification must be between "
				 "0 and %d\n", name, line, MAX_AMPLIFICATION);
			 return -2;
		  }
		bank->tone[i].amp=k;
	      }
	    else if (!strcmp(w[j], "note"))
	      {
		k=atoi(cp);
		if ((k<0 || k>127) || (*cp < '0' || *cp > '9'))
		  {
			 fprintf(stderr,
				 "%s: line %d: note must be between 0 and 127\n",
			    name, line);
		    return -2;
		  }
		bank->tone[i].note=k;
			}
		 else if (!strcmp(w[j], "pan"))
			{
		if (!strcmp(cp, "center"))
		  k=64;
		else if (!strcmp(cp, "left"))
		  k=0;
		else if (!strcmp(cp, "right"))
		  k=127;
		else
		  k=((atoi(cp)+100) * 100) / 157;
		if ((k<0 || k>127) ||
			 (k==0 && *cp!='-' && (*cp < '0' || *cp > '9')))
		  {
			 fprintf(stderr,
				 "%s: line %d: panning must be left, right, "
				 "center, or between -100 and 100\n",
				 name, line);
			 return -2;
		  }
		bank->tone[i].pan=k;
			}
		 else if (!strcmp(w[j], "keep"))
			{
		if (!strcmp(cp, "env"))
		  bank->tone[i].strip_envelope=0;
		else if (!strcmp(cp, "loop"))
		  bank->tone[i].strip_loop=0;
		else
		  {
			 fprintf(stderr, "%s: line %d: keep must be env or loop\n",
				 name, line);
			 return -2;
		  }
			}
		 else if (!strcmp(w[j], "strip"))
			{
		if (!strcmp(cp, "env"))
		  bank->tone[i].strip_envelope=1;
		else if (!strcmp(cp, "loop"))
		  bank->tone[i].strip_loop=1;
		else if (!strcmp(cp, "tail"))
		  bank->tone[i].strip_tail=1;
		else
		  {
			 fprintf(stderr, "%s: line %d: strip must be env, loop, or tail\n",
				 name, line);
			 return -2;
		  }
			}
		 else
			{
		fprintf(stderr, "%s: line %d: bad patch option %s\n",
			name, line, w[j]);
		return -2;
			}
	  }
		}
	 }
  if (ferror(fp))
	 {
		fprintf(stderr, "Can't read %s: %s\n", name, sys_errlist[errno]);
		close_file(fp);
		return -2;
	 }
  close_file(fp);
  return 0;
}
#ifdef __WIN32__
int __cdecl main(int argc, char **argv)
#else
int main(int argc, char **argv)
#endif
{
  int
	 c, cmderr=0, i, got_a_configuration=0, try_config_again=0,
	 need_stdin=0, need_stdout=0;

  int32
	 tmpi32, output_rate=DEFAULT_RATE;

  char
	 *output_name=0;

#if defined(AU_LINUX) || defined(AU_WIN32)
  int buffer_fragments=-1;
#endif

#ifdef __WIN32__
	int evil=0;
#endif

#ifdef DANGEROUS_RENICE
#include <sys/resource.h>
  int u_uid=getuid();
  if (setpriority(PRIO_PROCESS, 0, DANGEROUS_RENICE) < 0)
	 fprintf(stderr, "Couldn't set priority to %d.\n", DANGEROUS_RENICE);
  setreuid(u_uid, u_uid);
#endif

  if ((program_name=rindex(argv[0], '/'))) program_name++;
  else program_name=argv[0];
  if (argc==1)
	 {
		interesting_message();
		return 0;
	 }

  if (!read_config_file(CONFIG_FILE))
	 got_a_configuration=1;

  while ((c=getopt(argc, argv, "UI:P:L:c:A:C:ap:fo:O:s:Q:FD:hi:"
#if defined(AU_LINUX) || defined(AU_WIN32)
			"B:" /* buffer fragments */
#endif
#ifdef __WIN32__
			"e"	/* evil (be careful) */
#endif
			))>0)
    {
	 switch(c)
		{		case 'U': free_instruments_afterwards=1; break;
		case 'L': add_to_pathlist(optarg); try_config_again=1; break;
		case 'c':
	if (read_config_file(optarg)) cmderr++;
	else got_a_configuration=1;
	break;

		case 'Q':
	if (set_channel_flag(&quietchannels, atoi(optarg), "Quiet channel"))
	  cmderr++;
	break;

		case 'D':
	if (set_channel_flag(&drumchannels, atoi(optarg), "Drum channel"))
	  cmderr++;
	break;

		case 'O': /* output mode */
	if (set_play_mode(optarg))
	  cmderr++;
	break;

		case 'o':	output_name=optarg; break;

		case 'a': antialiasing_allowed=1; break;

		case 'f': fast_decay=(fast_decay) ? 0 : 1; break;

		case 'F': adjust_panning_immediately=1; break;

		case 's': /* sampling rate */
	i=atoi(optarg);
	if (i < 100) i *= 1000;
	if (set_value(&output_rate, i, MIN_OUTPUT_RATE, MAX_OUTPUT_RATE,
				"Resampling frequency")) cmderr++;
	break;

		case 'P': /* set overriding instrument */
	strncpy(def_instr_name, optarg, 255);
	def_instr_name[255]='\0';
	break;

		case 'I':
	if (set_value(&tmpi32, atoi(optarg), 0, 127,
				"Default program")) cmderr++;
	else default_program=tmpi32;
	break;
		case 'A':
	if (set_value(&amplification, atoi(optarg), 1, MAX_AMPLIFICATION,
				"Amplification")) cmderr++;
	break;
		case 'C':
	if (set_value(&control_ratio, atoi(optarg), 1, MAX_CONTROL_RATIO,
				"Control ratio")) cmderr++;
	break;
		case 'p':
	if (set_value(&tmpi32, atoi(optarg), 1, MAX_VOICES,
				"Polyphony")) cmderr++;
	else voices=tmpi32;
	break;

		case 'i':
	if (set_ctl(optarg))
	  cmderr++;
#if defined(AU_LINUX) || defined(AU_WIN32)
	else if (buffer_fragments==-1 && ctl->trace_playing)
	  /* user didn't specify anything, so use 2 for real-time response */
#if defined(AU_LINUX)
	  buffer_fragments=2;
#else
		buffer_fragments=3;		/* On Win32 2 is chunky */
#endif
#endif
	break;

#if defined(AU_LINUX) || defined(AU_WIN32)
		case 'B':
	if (set_value(&tmpi32, atoi(optarg), 0, 1000,
				"Buffer fragments")) cmderr++;
	else buffer_fragments=tmpi32;
	break;
#endif
#ifdef __WIN32__
		case 'e': /* evil */
			evil = 1;
			break;
#endif
		case 'h':
	help();
	return 0;

		default:
	cmderr++; break;
		}
	if(cmderr) break;

  if (!got_a_configuration)
	 {
		if (!try_config_again || read_config_file(CONFIG_FILE))
	cmderr++;
	 }

  /* If there were problems, give up now */
  if (cmderr || optind >= argc)
	 {
		fprintf(stderr, "Try %s -h for help\n", program_name);
		return 1; /* problems with command line */
	 }

  /* Set play mode parameters */
  play_mode->rate=output_rate;
  if (output_name)
	 {
		play_mode->name=output_name;
		if (!strcmp(output_name, "-"))
	need_stdout=1;
	 }
#if defined(AU_LINUX) || defined(AU_WIN32)
  if (buffer_fragments != -1)
	 play_mode->extra_param[0]=buffer_fragments;
#endif

  init_tables();

  if (optind<argc)
	 {
		int orig_optind=optind;

		while (optind<argc)
	if (!strcmp(argv[optind++], "-"))
	  need_stdin=1;
		optind=orig_optind;

		if (ctl->open(need_stdin, need_stdout))
	{
	  fprintf(stderr, "Couldn't open %s\n", ctl->id_name);
	  play_mode->close_output();
	  return 3;
	}
		/* Open output device */
		if (play_mode->open_output()<0)
	{
	  fprintf(stderr, "Couldn't open %s\n", play_mode->id_name);
	  ctl->close();
	  return 2;
	}

		if (!control_ratio)
	{
	  control_ratio = play_mode->rate / CONTROLS_PER_SECOND;
	  if(control_ratio<1)
		 control_ratio=1;
	  else if (control_ratio > MAX_CONTROL_RATIO)
		 control_ratio=MAX_CONTROL_RATIO;
	}

		if (*def_instr_name)
	set_default_instrument(def_instr_name);

#ifdef __WIN32__
		SetConsoleCtrlHandler (handler, TRUE);
		InitializeCriticalSection (&critSect);
		if(evil)
			if (!SetThreadPriority(GetCurrentThread(), THREAD_PRIORITY_ABOVE_NORMAL))
				fprintf(stderr, "Error raising process priority.\n");
#endif

		/* Return only when quitting */
		ctl->pass_playing_list(argc-optind, &argv[orig_optind]);

		play_mode->close_output();
		ctl->close();
#ifdef __WIN32__
			DeleteCriticalSection (&critSect);
#endif
		return 0;

	 }
  return 0;
}
